File: bin/orama-system/cidf/core/content_insertion_framework.py
"""
content_insertion_framework.py
──────────────────────────────
Platform-agnostic core library for the Content Insertion Decision Framework v1.3.
All other Python-based integrations (LangChain, CrewAI) import from here.

No external dependencies required.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Dict, List, Optional, Protocol


# ─── Data models ──────────────────────────────────────────────────────────────

@dataclass
class Task:
    """Describes what needs to be done."""
    task_type: str
    is_one_time: bool
    frequency_estimate: int
    content_static: bool
    requires_transformation: bool
    requires_conditional_logic: bool
    requires_external_integration: bool
    content_length_chars: int
    format_requirements: str
    signature: str
    estimated_setup_seconds: Optional[float] = None
    estimated_run_seconds: Optional[float] = None


@dataclass
class Env:
    """Describes what methods are available in the current environment."""
    field_accessible: bool
    editor_visible: bool
    paste_supported: bool
    upload_available: bool
    max_safe_chars_form_input: int = 10_000
    max_safe_chars_typing: int = 5_000
    formatting_preserved_on_paste: bool = True


@dataclass
class Decision:
    """Output of decide(). A blocked decision means notify the user; do not execute."""
    chosen_tool: Optional[str]
    fallback_chain: List[str]
    reason_codes: List[str]
    automation_justified: bool
    verification_required: bool = True
    blocked: bool = False
    notification_reason: Optional[str] = None


@dataclass
class AttemptLog:
    tool: str
    result: str
    detail: str = ""


@dataclass
class ExecutionResult:
    status: str
    tool: Optional[str] = None
    attempts: List[AttemptLog] = field(default_factory=list)
    notification_reason: Optional[str] = None


# ─── Core logic ───────────────────────────────────────────────────────────────

def automation_justified(task: Task) -> bool:
    """
    Return whether scripting may be considered after ranks 1–4 are exhausted.

    The gate fails closed for one-time static work and when known setup time
    exceeds one execution. Otherwise, any repeat/logic/transformation/external
    integration signal justifies considering scripting as the final fallback.
    """
    if task.is_one_time and task.content_static:
        return False
    if (
        task.estimated_setup_seconds is not None
        and task.estimated_run_seconds is not None
        and task.estimated_setup_seconds > task.estimated_run_seconds
    ):
        return False
    return (
        task.frequency_estimate >= 5
        or task.requires_conditional_logic
        or task.requires_transformation
        or task.requires_external_integration
    )


def decide(task: Task, env: Env) -> Decision:
    """
    Select the lowest-complexity eligible method.

    Scripting never displaces ranks 1–4. It is appended only as a last-resort
    fallback when the automation gate is open. If no method is eligible and the
    gate is closed, return an explicit blocked decision so the caller can notify
    the user instead of inventing an unavailable tool.
    """
    reasons: List[str] = []
    tools: List[str] = []

    if env.field_accessible and task.content_length_chars <= env.max_safe_chars_form_input:
        tools.append("direct_form_input")
    if env.editor_visible and task.content_length_chars <= env.max_safe_chars_typing:
        tools.append("direct_typing")
    if env.paste_supported:
        tools.append("clipboard_paste")
    if env.upload_available:
        tools.append("file_upload")

    justified = automation_justified(task)
    if tools:
        if justified:
            tools.append("scripting")
            reasons.append("scripting_deferred_until_lower_ranks_exhausted")
        chosen = tools[0]
        fallback = tools[1:]
        reasons.append(f"chosen_{chosen}")
        reasons.append(f"automation_justified={justified}")
        return Decision(
            chosen_tool=chosen,
            fallback_chain=fallback,
            reason_codes=reasons,
            automation_justified=justified,
        )

    if justified:
        reasons.extend(["chosen_scripting", "automation_justified=True"])
        return Decision(
            chosen_tool="scripting",
            fallback_chain=[],
            reason_codes=reasons,
            automation_justified=True,
        )

    notification_reason = "no_eligible_method_and_automation_gate_closed"
    reasons.extend([notification_reason, "automation_justified=False"])
    return Decision(
        chosen_tool=None,
        fallback_chain=[],
        reason_codes=reasons,
        automation_justified=False,
        blocked=True,
        notification_reason=notification_reason,
    )


# ─── Verification ─────────────────────────────────────────────────────────────

class Verifier(Protocol):
    """
    Implement this per environment (web/DOM, desktop, API).
    Never use visual confirmation as a substitute.
    """
    def refresh_once_if_needed(self) -> None: ...
    def extract_text(self) -> str: ...


def verify(verifier: Verifier, signature: str) -> bool:
    """
    Programmatic ground truth. Returns True only when a non-empty signature
    is confirmed present in the extracted text.
    """
    if not signature:
        raise ValueError("CIDF verification requires a non-empty signature.")
    verifier.refresh_once_if_needed()
    return signature in verifier.extract_text()


# ─── Execution loop ───────────────────────────────────────────────────────────

def execute_with_fallback(
    decision: Decision,
    executors: Dict[str, Callable[[str], None]],
    verifier: Verifier,
    content: str,
    signature: str,
) -> ExecutionResult:
    """
    Execute the chosen method, then each fallback in order.

    Every execution, verification, and registration outcome is logged. Executor
    or verifier errors move to the next eligible fallback; they never abort the
    chain or create a false success.
    """
    if not signature:
        raise ValueError("CIDF execution requires a non-empty signature.")

    if decision.blocked or decision.chosen_tool is None:
        return ExecutionResult(
            status="blocked",
            notification_reason=(
                decision.notification_reason
                or "no_eligible_method_and_automation_gate_closed"
            ),
        )

    chain = [decision.chosen_tool] + decision.fallback_chain
    attempts: List[AttemptLog] = []

    for tool in chain:
        executor = executors.get(tool)
        if executor is None:
            attempts.append(AttemptLog(tool=tool, result="no_executor_registered"))
            continue
        try:
            executor(content)
        except Exception as error:
            attempts.append(
                AttemptLog(
                    tool=tool,
                    result="execution_failed",
                    detail=type(error).__name__,
                )
            )
            continue
        try:
            verified = verify(verifier, signature)
        except Exception as error:
            attempts.append(
                AttemptLog(
                    tool=tool,
                    result="verification_error",
                    detail=type(error).__name__,
                )
            )
            continue
        if verified:
            attempts.append(AttemptLog(tool=tool, result="success"))
            return ExecutionResult(status="success", tool=tool, attempts=attempts)
        attempts.append(AttemptLog(tool=tool, result="verification_failed"))

    return ExecutionResult(
        status="failed",
        attempts=attempts,
        notification_reason="all_eligible_methods_exhausted",
    )

File: bin/orama-system/cidf/core/contentInsertionFramework.ts
/**
 * contentInsertionFramework.ts
 * ─────────────────────────────
 * Platform-agnostic TypeScript library for Content Insertion Decision Framework v1.3.
 * All TypeScript-based integrations (OpenClaw, etc.) import from here.
 *
 * No runtime dependencies required.
 */

export type TaskType = "content_insertion" | "automation" | "data_processing";
export type FormatRequirement = "plain" | "rich_text" | "strict_layout";
export type ToolName =
  | "direct_form_input"
  | "direct_typing"
  | "clipboard_paste"
  | "file_upload"
  | "scripting";

export interface Task {
  task_type: TaskType;
  is_one_time: boolean;
  frequency_estimate: number;
  content_static: boolean;
  requires_transformation: boolean;
  requires_conditional_logic: boolean;
  requires_external_integration: boolean;
  content_length_chars: number;
  format_requirements: FormatRequirement;
  signature: string;
  estimated_setup_seconds?: number | null;
  estimated_run_seconds?: number | null;
}

export interface Env {
  field_accessible: boolean;
  editor_visible: boolean;
  paste_supported: boolean;
  upload_available: boolean;
  max_safe_chars_form_input?: number;
  max_safe_chars_typing?: number;
  formatting_preserved_on_paste?: boolean;
}

export interface Decision {
  chosen_tool: ToolName | null;
  fallback_chain: ToolName[];
  reason_codes: string[];
  automation_justified: boolean;
  verification_required: true;
  blocked?: boolean;
  notification_reason?: string;
}

export interface Verifier {
  refreshOnceIfNeeded(): Promise<void>;
  extractText(): Promise<string>;
}

export interface AttemptLog {
  tool: ToolName;
  result:
    | "success"
    | "execution_failed"
    | "verification_failed"
    | "verification_error"
    | "no_executor_registered";
  detail?: string;
}

export interface ExecutionResult {
  status: "success" | "failed" | "blocked";
  tool?: ToolName;
  attempts: AttemptLog[];
  notification_reason?: string;
}

export function automationJustified(task: Task): boolean {
  if (task.is_one_time && task.content_static) return false;
  if (
    task.estimated_setup_seconds != null &&
    task.estimated_run_seconds != null &&
    task.estimated_setup_seconds > task.estimated_run_seconds
  ) {
    return false;
  }
  return (
    task.frequency_estimate >= 5 ||
    task.requires_conditional_logic ||
    task.requires_transformation ||
    task.requires_external_integration
  );
}

export function decide(task: Task, env: Env): Decision {
  const maxForm = env.max_safe_chars_form_input ?? 10_000;
  const maxTyping = env.max_safe_chars_typing ?? 5_000;

  const eligibleRanked: ToolName[] = [
    ...(env.field_accessible && task.content_length_chars <= maxForm
      ? (["direct_form_input"] as const)
      : []),
    ...(env.editor_visible && task.content_length_chars <= maxTyping
      ? (["direct_typing"] as const)
      : []),
    ...(env.paste_supported ? (["clipboard_paste"] as const) : []),
    ...(env.upload_available ? (["file_upload"] as const) : []),
  ];

  const justified = automationJustified(task);
  if (eligibleRanked.length > 0) {
    const tools: ToolName[] = justified
      ? [...eligibleRanked, "scripting"]
      : eligibleRanked;
    const chosen = tools[0];
    if (chosen === undefined) {
      // Unreachable given eligibleRanked.length > 0 above, but
      // noUncheckedIndexedAccess correctly can't prove that statically --
      // fail loudly rather than silently return an invalid Decision.
      throw new Error("CIDF internal error: eligible tools list was unexpectedly empty");
    }
    const reasons: string[] = [
      ...(justified ? ["scripting_deferred_until_lower_ranks_exhausted"] : []),
      "chosen_" + chosen,
      "automation_justified=" + justified,
    ];
    return {
      chosen_tool: chosen,
      fallback_chain: tools.slice(1),
      reason_codes: reasons,
      automation_justified: justified,
      verification_required: true,
    };
  }

  if (justified) {
    return {
      chosen_tool: "scripting",
      fallback_chain: [],
      reason_codes: ["chosen_scripting", "automation_justified=True"],
      automation_justified: true,
      verification_required: true,
    };
  }

  const notificationReason = "no_eligible_method_and_automation_gate_closed";
  return {
    chosen_tool: null,
    fallback_chain: [],
    reason_codes: [notificationReason, "automation_justified=False"],
    automation_justified: false,
    verification_required: true,
    blocked: true,
    notification_reason: notificationReason,
  };
}

export async function verify(verifier: Verifier, signature: string): Promise<boolean> {
  if (!signature) {
    throw new Error("CIDF verification requires a non-empty signature.");
  }
  await verifier.refreshOnceIfNeeded();
  const text = await verifier.extractText();
  return text.includes(signature);
}

export async function executeWithFallback(
  decision: Decision,
  executors: Partial<Record<ToolName, (content: string) => Promise<void>>>,
  verifier: Verifier,
  content: string,
  signature: string,
): Promise<ExecutionResult> {
  if (!signature) {
    throw new Error("CIDF execution requires a non-empty signature.");
  }
  if (decision.blocked || !decision.chosen_tool) {
    return {
      status: "blocked",
      attempts: [],
      notification_reason:
        decision.notification_reason ||
        "no_eligible_method_and_automation_gate_closed",
    };
  }

  const chain: ToolName[] = [decision.chosen_tool, ...decision.fallback_chain];
  let attempts: AttemptLog[] = [];

  for (const tool of chain) {
    const executor = executors[tool];
    if (!executor) {
      attempts = [...attempts, { tool, result: "no_executor_registered" }];
      continue;
    }
    try {
      await executor(content);
    } catch (error) {
      attempts = [
        ...attempts,
        {
          tool,
          result: "execution_failed",
          detail: error instanceof Error ? error.name : "UnknownError",
        },
      ];
      continue;
    }
    try {
      const verified = await verify(verifier, signature);
      if (verified) {
        attempts = [...attempts, { tool, result: "success" }];
        return { status: "success", tool, attempts };
      }
    } catch (error) {
      attempts = [
        ...attempts,
        {
          tool,
          result: "verification_error",
          detail: error instanceof Error ? error.name : "UnknownError",
        },
      ];
      continue;
    }
    attempts = [...attempts, { tool, result: "verification_failed" }];
  }

  return {
    status: "failed",
    attempts,
    notification_reason: "all_eligible_methods_exhausted",
  };
}
